man/man1/lgt2rst.1
.TH lgt2rst 1 "April 2, 2025" "Logtalk 3.91.0" "Logtalk Documentation"

.SH NAME
lgt2rst \- Logtalk script for converting XML documenting files into reStructuredText files, notably for use with Sphinx

.SH SYNOPSIS
.B lgt2rst
[\fB-d \fIdirectory\fR]
[\fB-i \fIindex\fR]
[\fB-t \fItitle\fR]
[\fB-p \fIprocessor\fR]
[\fB-s\fI\fR]
[\fB-m\fI\fR]
[\fB-l \fImapping\fR]
.br
.B lgt2rst
.B \-h

.SH DESCRIPTION
\fBlgt2rst\fR is a Logtalk script for converting all Logtalk XML documenting files in the current directory into reStructuredText files, notably for use with Sphinx. Assumes UTF-8 encoding.

.SH OPTIONS
.TP
.B \-d
.I directory
Output directory for the reStructuredText files.
.TP
.B \-i
.I index
Name of the index file.
.TP
.B \-t
.I title
Title to be used in the index file.
.TP
.B \-p
.I processor
XSLT processor.
.TP
.B \-s
Runs sphinx-quickstart script.
.TP
.B \-m
Runs make html (requires also using the \fB-s\fI\fR option).
.TP
.B \-l
.I mapping
Name of the Intersphinx mapping for linking library APIs back to library descriptions.
.TP
.B \-h
Prints a summary of the script options.

.SH REQUIREMENTS
Requires a compatible XSLT processor. See the script help screen for more information.
.PP
The use of the \fB-s\fR, \fB-m\fR, and \fB-l\fR options require Sphinx, Intersphinx extension, Read the Docs Sphinx theme, and Pygments to be installed.
.PP
The use of the \fB-l\fR option requires that the used \fBconf.py\fR file defines the mapping. The library description files must have a reStructuredText target in the first line with the name of the library prefixed by \fBlibrary_\fR. See the Intersphinx documentation for details.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.dtd
The Logtalk entity DTD file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.dtd
The Logtalk index DTD file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/custom.ent
The Logtalk custom entities file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.xsd
The Logtalk entity XML Schema file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.xsd
The Logtalk index XML Schema file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_entity_to_rst.xsl
The Logtalk entity XSLT file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_index_to_rst.xsl
The Logtalk index XSLT file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/conf.py.jinja
The conf.py template file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/css/sphinx/custom.css
The CSS file used to customize the Sphinx Read the Docs theme.

.SH "SEE ALSO"
The Sphinx website at https://www.sphinx-doc.org/
.PP
The Pygments website at https://pygments.org/
.PP
The Read the Docs theme website at https://github.com/readthedocs/sphinx_rtd_theme
.PP
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
logtalk_user_setup(1),\ lgt2html(1),\ lgt2xml(1),\ lgt2pdf(1),\ lgt2md(1),\ lgt2txt(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2025 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2025 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.91.0 - April 8, 2025
======================

Logtalk compiler and runtime
----------------------------

* FIXED: The documentation of the `logtalk::expand_library_path/2` predicate
to clarify that only expands library aliases (atoms) or compound terms in
library notation.

Library
-------

* FIXED: Avoid a non-tail recursive predicate linter warning when loading the
`random` library with the `tail_recursive` set to `warning`.

Tools
-----

* CHANGED: The `lgtunit` tool messages for passed and skipped tests to also
include the flaky flag.

* ADDED: Flaky test property/note support to the `lgtunit` tool TAP and xUnit
output formats.

* IMPROVED: The `lgtdoc` tool no longer replaces the ASCII comma with the
Unicode fullwidth comma when generating reStructuredText index entries for
entity names.

* UPDATED: Developer tools requirements for Ubuntu to avoid requiring the
full Texlive installation.

* FIXED: The `lgtdoc` tool to ignore the `home` and `startup` library names
when looking for the name of the library when processing directory files.

* FIXED: Renamed the `lgtdoc` tool Sphinx configuration template file to
`conf.py.jinja` as the still documented `conf.py_t` name is no longer
recognized by recent Sphinx versions.

* FIXED: Bug in the `lgtunit` tool where the `quick_check/3` test dialect
would throw an error when using the `pc/1` option.

* FIXED: Bug in the `lgtunit` tool where broken QuickCheck tests would be
reported as backend bugs.

* FIXED: Regression in the `lgtunit` tool TAP and xUnit output formats where
failed tests would not be reported.

* FIXED: Missing handling of broken QuickCheck tests in the `lgtunit` tool
TAP and xUnit output formats.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* UPDATED: VSCode support for the changes to the `lgtunit` tool messages.

Installers and installation scripts
-----------------------------------

* UPDATED: The Windows installation script to also detect the new SICStus
Prolog 4.10.x versions.

* UPDATED: The portable Docker image documentation on how to browse Allure
reports from the host.

* UPDATED: The portable Docker image to also include the `issue_creator`
tool dependencies `gh` and `glab`.

* FIXED: The portable Docker image Texlive requirements.

* FIXED: The portable Docker image Allure installation.


3.90.1 - March 27, 2025
=======================

Tools
-----

* FIXED: PowerShell script coding patterns that could trigger false positives
when using some vendors security scanning tools.


3.90.0 - March 25, 2025
=======================

Prolog adapter and integration files
------------------------------------

* CHANGED: Moved all common setup code from the Bash shell integration scripts
to a shared setup script.

* CHANGED: Moved all common setup code from the PowerShell integration scripts
to a module.

* FIXED: The PowerShell integration scripts were not setting the `LOGTALKUSER`
environment variable correctly when not defined by the user.

* FIXED: The `gplgt.ps1` PowerShell integration script to run without the GUI.

Documentation
-------------

* ADDED: Support for generating single file versions of the Handbook and APIs
documentation in LLM friendly Markdown format.

* IMPROVED: The documentation of the `logtalk_tester.sh` Bash script.

Library
-------

* ADDED: Predicates `between/4` and `sequence/4-5` to the `float` object in
the `types` library.

* ADDED: Predicates `softmax/2-3` to the `numberlist` object in the `types`
library.

* ADDED: Predicates to the `random` library for drawing samples from common
probability distributions.

Tools
-----

* CHANGED: Moved all common setup code from the `lgtdoc` tool Bash shell
scripts to a shared setup script.

* CHANGED: Moved all common setup code from the `lgtdoc` tool PowerShell
integration scripts to a module.

* CHANGED: The `packs` tool `lgtenv.*` scripts to exit with a specific code
for each error.

* ADDED: Predicates `write_max_depth/1` and `set_write_max_depth/1` to the
`debugger` tool.

* IMPROVED: The `logtalk_tester.*` scripts handling of the `-b` and `-u`
options.

* IMPROVED: The `lgt2svg.*` scripts handling of file paths and checking of the
d2 and Graphviz commands availability.

* FIXED: The `logtalk_tester.ps1` PowerShell script `xunit_net_v2` format
option.

* FIXED: The `issue_creator` tool to workaround GitLab creating CI/CD pipelines
in a detached HEAD state.

* FIXED: The `logtalk_allure_report.ps1` PowerShell script creation of the
`executor.json` file to wrap the key-value pairs in object syntax.

* FIXED: The `logtalk_doclet.ps1` PowerShell script handling of the prefix and
timeout options.

* FIXED: The Bash shell scripts to exit with code 1 in case of wrong usage.

Examples
--------

* ADDED: New `apache_poi` Java example of using the Apache POI library to read
and write Excel spreadsheets.

* ADDED: New `lips` example of the classical naive list reverse benchmark
computing million of logical inferences per second.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------


tools/lgtdoc/xml/lgt2rst.ps1
#############################################################################
##
##   XML documenting files to reStructuredText files conversion script
##   Last updated on March 23, 2025
##
##   This file is part of Logtalk <https://logtalk.org/>
##   Copyright 2022-2025 Paulo Moura <pmoura@logtalk.org>
##   Copyright 2022 Hans N. Beck
##   SPDX-License-Identifier: Apache-2.0
##
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##
##       http://www.apache.org/licenses/LICENSE-2.0
##
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
##
#############################################################################


#Requires -Version 7.3

[CmdletBinding()]
param(
	[Parameter()]
	[String]$d = $pwd,
	[String]$i = "index.rst",
	[String]$t = "Documentation index",
	[String]$l = "",
	[Switch]$s,
	[Switch]$m,
	[Switch]$v,
	[Switch]$h
)

function Write-Script-Version {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve
	Write-Output "$myName 5.2"
}

function Write-Usage-Help() {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path "$myFullName" -leaf -Resolve

	Write-Output "This script converts all Logtalk XML documenting files in the"
	Write-Output "current directory to reStructuredText files for use with Sphinx"
	Write-Output ""
	Write-Output "Usage:"
	Write-Output "  $myName [-d directory] [-i index] [-t title] [-s] [-m] [-l mapping]"
	Write-Output "  $myName -v"
	Write-Output "  $myName -h"
	Write-Output ""
	Write-Output "Optional arguments:"
	Write-Output "  -d output directory for the generated files (default is $d)"
	Write-Output "  -i name of the index file (default is $i)"
	Write-Output "  -t title to be used in the index file (default is `"$t`")"
	Write-Output "  -s run sphinx-quickstart script"
	Write-Output "  -m run make html (requires also using the -s option)"
	Write-Output "  -l Intersphinx mapping for linking library APIs to library descriptions (requires -s option)"
	Write-Output "  -v print version"
	Write-Output "  -h help"
	Write-Output ""
}

function Confirm-Parameters() {

	if (-not(Test-Path $d)) { # cannot be ""
		Write-Error "The $d output directory does not exist!"
		Start-Sleep -Seconds 2
		Exit 1
	}

	if ($v -eq $true) {
		Write-Script-Version
		Exit 0
	}

	if ($h -eq $true) {
		Write-Usage-Help
		Exit 0
	}

}

function New-Index-File() {
	New-Item -Path . -Name "$i" -ItemType "file" -Force > $null

	Add-Content -Path "$i" -Value ""
	Add-Content -Path "$i" -Value $t
	$underline = "=" * $t.Length
	Add-Content -Path "$i" -Value $underline
	Add-Content -Path "$i" -Value ""
	Add-Content -Path "$i" -Value ""

	if (Test-Path "directory_index.xml") {
		Add-Content -Path "$i" -Value ".. toctree::"
		Add-Content -Path "$i" -Value "   :maxdepth: 3"
		Add-Content -Path "$i" -Value "   :caption: Contents"
		Add-Content -Path "$i" -Value ""
		Add-Content -Path "$i" -Value "   library_index"
		Add-Content -Path "$i" -Value "   directory_index"
		Add-Content -Path "$i" -Value "   entity_index"
		Add-Content -Path "$i" -Value "   predicate_index"
		Add-Content -Path "$i" -Value ""
		Add-Content -Path "$i" -Value "Indices and tables"
		Add-Content -Path "$i" -Value "=================="
		Add-Content -Path "$i" -Value ""
		Add-Content -Path "$i" -Value '* :ref:`genindex`'
		Add-Content -Path "$i" -Value '* :ref:`search`'
	} elseif (Get-ChildItem -Path . -Filter .\*.xml | Select-String -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
		Add-Content -Path "$i" -Value ".. toctree::"
		Add-Content -Path "$i" -Value "   :maxdepth: 1"
		Add-Content -Path "$i" -Value ""
		Get-ChildItem -Path . -Filter .\*.xml |
		Foreach-Object {
			if ($_ | Select-String -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
				$entity = ($_.BaseName -replace '_[^_]*$')
				$pars   = ($_.BaseName -replace '.*_')
				Write-Output "  indexing $($_.BaseName).rst"
				if ($pars -ne "" -and [int]$pars -gt 0) {
					Add-Content -Path "$i" -Value "   $entity/$pars <$($_.BaseName)>"
				} else {
					Add-Content -Path "$i" -Value "   $entity <$($_.BaseName)>"
				}
			}
		}
	}

	$date = Get-Date -Format "ddd MMM dd HH:mm:ss K yyyy"

	Add-Content -Path "$i" -Value ""
	Add-Content -Path "$i" -Value "Generated on $date"
}

###################### here it starts ############################

Import-Module (Join-Path $PSScriptRoot "LogtalkSetupEnv.psm1")
Initialize-LogtalkEnvironment

Confirm-Parameters

$entity_xslt = "$env:LOGTALKUSER/tools/lgtdoc/xml/logtalk_entity_to_rst.xsl"
$index_xslt  = "$env:LOGTALKUSER/tools/lgtdoc/xml/logtalk_index_to_rst.xsl"

if (!(Test-Path "logtalk_entity.dtd")) {
	Copy-Item -Path "$env:LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.dtd" -Destination .
}

if (!(Test-Path "logtalk_index.dtd")) {
	Copy-Item -Path "$env:LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.dtd" -Destination .
}

if (!(Test-Path "custom.ent")) {
	Copy-Item -Path "$env:LOGTALKUSER/tools/lgtdoc/xml/custom.ent" -Destination .
}

if (!(Test-Path "logtalk_entity.xsd")) {
	Copy-Item -Path "$env:LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.xsd" -Destination .
}

if (!(Test-Path "logtalk_index.xsd")) {
	Copy-Item -Path "$env:LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.xsd" -Destination .
}

if (Select-String -Path .\*.xml -Pattern '<logtalk' -CaseSensitive -SimpleMatch -Quiet) {
	Write-Output "Converting XML files to reStructuredText files..."

	$xslt_settings = New-Object System.Xml.Xsl.XsltSettings
	$xslt_settings.EnableDocumentFunction = $true

	$xml_reader_settings = New-Object System.Xml.XmlReaderSettings
	$xml_reader_settings.DtdProcessing = [System.Xml.DtdProcessing]::Ignore

	$xml_url_resolver = New-Object System.Xml.XmlUrlResolver

	$entity_xslt_object = New-Object System.Xml.Xsl.XslCompiledTransform
	$entity_xslt_object.Load($entity_xslt, $xslt_settings, $xml_url_resolver)

	$index_xslt_object = New-Object System.Xml.Xsl.XslCompiledTransform;
	$index_xslt_object.Load($index_xslt, $xslt_settings, $xml_url_resolver)

	if ($l) {
		$xslArguments = New-Object System.Xml.Xsl.XsltArgumentList
		$xslArguments.AddParam("mapping", [string]::Empty, $l)
	} else {
		$xslArguments = $null;
	}

	Get-ChildItem -Path . -Filter *.xml |
	Foreach-Object {
		if (Select-String -Path $_ -Pattern '<logtalk_entity' -CaseSensitive -SimpleMatch -Quiet) {
			Write-Output "  converting $($_.Name)"
			$file = Join-Path "$pwd" "$($_.Name)"
			$rst = Join-Path $d "$($_.BaseName).rst"
			$reader = [System.Xml.XmlReader]::Create($file, $xml_reader_settings)
			$fs = New-Object IO.FileStream $rst, 'Append', 'Write', 'Read'
			$writer = New-Object System.IO.StreamWriter($fs)
			$entity_xslt_object.Transform($reader, $xslArguments, $writer)
			$writer.Close()
			$reader.Dispose()
		}
	}
	Get-ChildItem -Path . -Filter .\*.xml |
	Foreach-Object {
		if (Select-String -Path $_ -Pattern '<logtalk_index' -CaseSensitive -SimpleMatch -Quiet) {
			Write-Output "  converting $($_.Name)"
			$file = Join-Path "$pwd" "$($_.Name)"
			$rst = Join-Path "$d" "$($_.BaseName).rst"
			$reader = [System.Xml.XmlReader]::Create($file, $xml_reader_settings)
			$fs = New-Object IO.FileStream $rst, 'Append', 'Write', 'Read'
			$writer = New-Object System.IO.StreamWriter($fs)
			$index_xslt_object.Transform($reader, $xslArguments, $writer)
			$writer.Close()
			$reader.Dispose()
		}
	}
	Write-Output "conversion done"
	Write-Output ""
	Write-Output "generating $i file..."
	New-Index-File
	Write-Output "$i file generated"
	Write-Output ""
	if ($s -eq $true) {
		Move-Item -Path index.rst -Destination index.rst.backup -Force
		sphinx-quickstart --templatedir="$env:LOGTALKUSER/tools/lgtdoc/xml"
		if (-not (Test-Path _static/css)) {
			New-Item -Path _static/css -ItemType Directory > $null
		}
		Copy-Item -Path "$env:LOGTALKUSER/tools/lgtdoc/xml/css/sphinx/custom.css" -Destination _static/css/custom.css
		Move-Item -Path index.rst.backup -Destination index.rst -Force
	}
	if ($s -eq $true -and $m -eq $true) {
		.\make html
	}
} else {
	Write-Output ""
	Write-Output "No XML files exist in the current directory!"
	Write-Output ""
}
